src/d3d12.zig

```diff
@@ -553,22 +549,442 @@ pub const Device = struct {
+                c.D3D12_HEAP_TYPE_UPLOAD => .gpu_to_cpu,
+                c.D3D12_HEAP_TYPE_READBACK => .cpu_to_gpu,
+                else => .gpu_only,
+            },
+            .resource_desc = &resource_desc,
+            .clear_value = null,
+            .resource_category = .buffer,
+            .initial_state = initial_state,
+        };
+
+        return try device.mem_allocator.createResource(&create_desc);
+    }
+};
+
+pub const ResourceCategory = enum {
+    buffer,
+    rtv_dsv_texture,
+    other_texture,
+
+    pub inline fn heapUsable(self: ResourceCategory, heap: HeapCategory) bool {
+        return switch (heap) {
+            .all => true,
+            .buffer => self == .buffer,
+            .rtv_dsv_texture => self == .rtv_dsv_texture,
+            .other_texture => self == .other_texture,
+        };
+    }
+};
+
+pub const HeapCategory = enum {
+    all,
+    buffer,
+    rtv_dsv_texture,
+    other_texture,
+};
+
+pub const AllocationCreateDescriptor = struct {
+    location: gpu_allocator.MemoryLocation,
+    size: u64,
+    alignment: u64,
+    resource_category: ResourceCategory,
+};
+
+pub const ResourceCreateDescriptor = struct {
+    location: gpu_allocator.MemoryLocation,
+    resource_category: ResourceCategory,
+    resource_desc: *const c.D3D12_RESOURCE_DESC,
+    clear_value: ?*const c.D3D12_CLEAR_VALUE,
+    initial_state: c.D3D12_RESOURCE_STATES,
+};
+
+/// Stores a group of heaps
+pub const MemoryAllocator = struct {
+    const max_memory_groups = 9;
+    device: *Device,
+
+    memory_groups: std.BoundedArray(MemoryGroup, max_memory_groups),
+    allocation_sizes: gpu_allocator.AllocationSizes,
+
+    /// a single heap,
+    /// use the gpu_allocator field to allocate chunks of memory
+    pub const MemoryHeap = struct {
+        index: usize,
+        heap: *c.ID3D12Heap,
+        size: u64,
+        gpu_allocator: gpu_allocator.Allocator,
+
+        pub fn init(
+            group: *MemoryGroup,
+            index: usize,
+            size: u64,
+            dedicated: bool,
+        ) gpu_allocator.Error!MemoryHeap {
+            const heap = blk: {
+                var desc = c.D3D12_HEAP_DESC{
+                    .SizeInBytes = size,
+                    .Properties = group.heap_properties,
+                    .Alignment = @intCast(c.D3D12_DEFAULT_MSAA_RESOURCE_PLACEMENT_ALIGNMENT),
+                    .Flags = switch (group.heap_category) {
+                        .all => c.D3D12_HEAP_FLAG_NONE,
+                        .buffer => c.D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS,
+                        .rtv_dsv_texture => c.D3D12_HEAP_FLAG_ALLOW_ONLY_RT_DS_TEXTURES,
+                        .other_texture => c.D3D12_HEAP_FLAG_ALLOW_ONLY_NON_RT_DS_TEXTURES,
+                    },
+                };
+
+                var heap: ?*c.ID3D12Heap = null;
+                const hr = group.owning_pool.device.d3d_device.lpVtbl.*.CreateHeap.?(
+                    group.owning_pool.device.d3d_device,
```

Comment:
Typically I'm binding a local variable for D3D12 method calls to make it easier to keep the 1st arg in sync.